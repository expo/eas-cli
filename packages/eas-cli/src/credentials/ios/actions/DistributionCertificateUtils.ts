import assert from 'assert';
import chalk from 'chalk';
import dateformat from 'dateformat';

import { AccountFragment, AppleDistributionCertificateFragment } from '../../../graphql/generated';
import Log, { learnMore } from '../../../log';
import { promptAsync } from '../../../prompts';
import { fromNow } from '../../../utils/date';
import { CredentialsContext } from '../../context';
import { askForUserProvidedAsync } from '../../utils/promptForCredentials';
import { AppLookupParams } from '../api/graphql/types/AppLookupParams';
import {
  DistributionCertificate,
  DistributionCertificateStoreInfo,
} from '../appstore/Credentials.types';
import { filterRevokedDistributionCertsFromEasServers } from '../appstore/CredentialsUtils';
import { AppleTooManyCertsError } from '../appstore/distributionCertificate';
import { distributionCertificateSchema } from '../credentials';
import { validateDistributionCertificateAsync } from '../validators/validateDistributionCertificate';
import { formatAppleTeam } from './AppleTeamFormatting';

export function formatDistributionCertificate(
  distributionCertificate: AppleDistributionCertificateFragment,
  validSerialNumbers?: string[]
): string {
  const {
    serialNumber,
    developerPortalIdentifier,
    appleTeam,
    validityNotBefore,
    validityNotAfter,
    updatedAt,
  } = distributionCertificate;
  let line: string = '';
  if (developerPortalIdentifier) {
    line += `Cert ID: ${developerPortalIdentifier}`;
  }
  line += `${line === '' ? '' : ', '}Serial number: ${serialNumber}${
    appleTeam ? `, ${formatAppleTeam(appleTeam)}` : ''
  }`;
  line += chalk.gray(
    `\n    Created: ${fromNow(new Date(validityNotBefore))} ago, Updated: ${fromNow(
      new Date(updatedAt)
    )} ago,`
  );
  line += chalk.gray(`\n    Expires: ${dateformat(validityNotAfter, 'expiresHeaderFormat')}`);
  const apps = distributionCertificate.iosAppBuildCredentialsList.map(
    buildCredentials => buildCredentials.iosAppCredentials.app
  );
  if (apps.length) {
    const appFullNames = apps.map(app => app.fullName).join(',');
    line += chalk.gray(`\n    ðŸ“² Used by: ${appFullNames}`);
  }

  if (validSerialNumbers?.includes(serialNumber)) {
    line += chalk.gray("\n    âœ… Currently valid on Apple's servers.");
  } else {
    line += '';
  }
  return line;
}

async function selectDistributionCertificateAsync(
  distCerts: AppleDistributionCertificateFragment[],
  validDistributionCertificates?: AppleDistributionCertificateFragment[]
): Promise<AppleDistributionCertificateFragment | null> {
  const validDistCertSerialNumbers = validDistributionCertificates?.map(
    distCert => distCert.serialNumber
  );
  const { chosenDistCert } = await promptAsync({
    type: 'select',
    name: 'chosenDistCert',
    message: 'Select certificate from the list.',
    choices: distCerts.map(distCert => ({
      title: formatDistributionCertificate(distCert, validDistCertSerialNumbers),
      value: distCert,
    })),
  });
  return chosenDistCert;
}

/**
 * select a distribution certificate from an account (validity status shown on a best effort basis)
 * */
export async function selectDistributionCertificateWithDependenciesAsync(
  ctx: CredentialsContext,
  account: AccountFragment
): Promise<AppleDistributionCertificateFragment | null> {
  const distCertsForAccount = await ctx.ios.getDistributionCertificatesForAccountAsync(account);
  if (distCertsForAccount.length === 0) {
    Log.warn(`There are no Distribution Certificates available in your EAS account.`);
    return null;
  }
  if (!ctx.appStore.authCtx) {
    return selectDistributionCertificateAsync(distCertsForAccount);
  }

  // get valid certs on the developer portal
  const certInfoFromApple = await ctx.appStore.listDistributionCertificatesAsync();
  const validDistCerts = await filterRevokedDistributionCertsFromEasServers(
    distCertsForAccount,
    certInfoFromApple
  );

  return selectDistributionCertificateAsync(distCertsForAccount, validDistCerts);
}

/**
 * select a distribution certificate from a valid set (curated on a best effort basis)
 * */
export async function selectValidDistributionCertificateAsync(
  ctx: CredentialsContext,
  appLookupParams: AppLookupParams
): Promise<AppleDistributionCertificateFragment | null> {
  const distCertsForAccount = await ctx.ios.getDistributionCertificatesForAccountAsync(
    appLookupParams.account
  );
  if (distCertsForAccount.length === 0) {
    Log.warn(`There are no Distribution Certificates available in your EAS account.`);
    return null;
  }
  if (!ctx.appStore.authCtx) {
    return selectDistributionCertificateAsync(distCertsForAccount);
  }

  // filter by apple team
  assert(ctx.appStore.authCtx, 'authentication to the Apple App Store is required');
  const appleTeamIdentifier = ctx.appStore.authCtx.team.id;
  const distCertsForAppleTeam = distCertsForAccount.filter(distCert => {
    return !distCert.appleTeam || distCert.appleTeam.appleTeamIdentifier === appleTeamIdentifier;
  });

  // filter by valid certs on the developer portal
  const certInfoFromApple = await ctx.appStore.listDistributionCertificatesAsync();
  const validDistCerts = filterRevokedDistributionCertsFromEasServers(
    distCertsForAppleTeam,
    certInfoFromApple
  );
  Log.log(
    `${validDistCerts.length}/${distCertsForAccount.length} Distribution Certificates are currently valid for Apple Team ${ctx.appStore.authCtx?.team.id}.`
  );
  return selectDistributionCertificateAsync(validDistCerts);
}

const APPLE_DIST_CERTS_TOO_MANY_GENERATED_ERROR = `
You can have only ${chalk.underline(
  'three'
)} Apple Distribution Certificates generated on your Apple Developer account.
Revoke the old ones or reuse existing from your other apps.
Remember that Apple Distribution Certificates are not application specific!
`;

export async function provideOrGenerateDistributionCertificateAsync(
  ctx: CredentialsContext
): Promise<DistributionCertificate> {
  if (!ctx.nonInteractive) {
    const userProvided = await promptForDistCertAsync(ctx);
    if (userProvided) {
      if (!ctx.appStore.authCtx) {
        Log.warn(
          'Unable to validate distribution certificate, you are not authenticated with Apple.'
        );
        return userProvided;
      } else {
        const isValid = await validateDistributionCertificateAsync(ctx, userProvided);
        if (!isValid) {
          Log.warn("Provided Distribution Certificate is no longer valid on Apple's server");
        }
        return isValid ? userProvided : await provideOrGenerateDistributionCertificateAsync(ctx);
      }
    }
  }
  return await generateDistributionCertificateAsync(ctx);
}

async function promptForDistCertAsync(
  ctx: CredentialsContext
): Promise<DistributionCertificate | null> {
  let initialValues: { teamId?: string } = {};
  if (ctx.appStore.authCtx) {
    initialValues = {
      teamId: ctx.appStore.authCtx.team.id,
    };
  }
  const userProvided = await askForUserProvidedAsync(distributionCertificateSchema, initialValues);
  if (!userProvided) {
    return null;
  }
  if (ctx.appStore.authCtx && userProvided.teamId === initialValues.teamId) {
    return {
      ...userProvided,
      teamName: ctx.appStore.authCtx.team.name,
    };
  }
  return userProvided;
}

async function generateDistributionCertificateAsync(
  ctx: CredentialsContext
): Promise<DistributionCertificate> {
  await ctx.appStore.ensureAuthenticatedAsync();
  try {
    return await ctx.appStore.createDistributionCertificateAsync();
  } catch (e) {
    if (e instanceof AppleTooManyCertsError) {
      const distCerts = await ctx.appStore.listDistributionCertificatesAsync();
      Log.warn('Maximum number of Distribution Certificates generated on Apple Developer Portal.');
      Log.warn(APPLE_DIST_CERTS_TOO_MANY_GENERATED_ERROR);

      if (ctx.nonInteractive) {
        throw new Error(
          "Start the CLI without the '--non-interactive' flag to revoke existing certificates."
        );
      }

      Log.log(
        chalk.grey(
          `âœ…  Distribution Certificates can be revoked with no side effects for App Store builds.`
        )
      );
      Log.log(learnMore('https://docs.expo.dev/distribution/app-signing/#summary'));
      Log.newLine();

      const { distCertsToRevoke } = await promptAsync({
        type: 'multiselect',
        name: 'distCertsToRevoke',
        message: 'Select certificates to revoke.',
        // @ts-expect-error property missing from `@types/prompts`
        optionsPerPage: 20,
        choices: distCerts.map(distCert => ({
          value: distCert,
          title: formatDistributionCertificateFromApple(distCert),
        })),
      });

      if (distCertsToRevoke.length > 0) {
        const ids = distCertsToRevoke.map(({ id }: DistributionCertificateStoreInfo) => id);
        await ctx.appStore.revokeDistributionCertificateAsync(ids);
      }
    } else {
      throw e;
    }
  }
  return await generateDistributionCertificateAsync(ctx);
}

function formatDistributionCertificateFromApple(
  appleInfo: DistributionCertificateStoreInfo
): string {
  const { name, status, id, expires, created, ownerName, serialNumber } = appleInfo;
  const expiresDate = new Date(expires * 1000).toDateString();
  const createdDate = new Date(created * 1000).toDateString();
  return `${name} (${status}) - Cert ID: ${id}, Serial number: ${serialNumber}, Team ID: ${appleInfo.ownerId}, Team name: ${ownerName}
    expires: ${expiresDate}, created: ${createdDate}`;
}
