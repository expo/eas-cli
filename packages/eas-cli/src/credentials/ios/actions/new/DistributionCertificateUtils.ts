import { AppleDistributionCertificate, AppleTeam } from '../../../../graphql/generated';

export function formatDistributionCertificate({
  developerPortalIdentifier,
  serialNumber,
  appleTeam,
  validityNotBefore,
  validityNotAfter,
}: AppleDistributionCertificate): string {
  let line: string = '';
  if (developerPortalIdentifier) {
    line += `Cert ID: ${developerPortalIdentifier}`;
  }
  line += `${line === '' ? '' : ', '}Serial number: ${serialNumber}${
    appleTeam ? `, ${formatAppleTeam(appleTeam)}` : ''
  }, Created: ${validityNotBefore}, Expires: ${validityNotAfter}`;
  return line;
}

function formatAppleTeam({ appleTeamIdentifier, appleTeamName }: AppleTeam): string {
  return `Team ID: ${appleTeamIdentifier}${appleTeamName ? `, Team name: ${appleTeamName}` : ''}`;
}
