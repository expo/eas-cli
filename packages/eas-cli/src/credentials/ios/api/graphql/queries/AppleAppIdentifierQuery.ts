import { print } from 'graphql';
import gql from 'graphql-tag';

import { ExpoGraphqlClient } from '../../../../../commandUtils/context/contextUtils/createGraphqlClient';
import { withErrorHandlingAsync } from '../../../../../graphql/client';
import {
  AppleAppIdentifierByBundleIdQuery,
  AppleAppIdentifierFragment,
} from '../../../../../graphql/generated';
import { AppleAppIdentifierFragmentNode } from '../../../../../graphql/types/credentials/AppleAppIdentifier';

export const AppleAppIdentifierQuery = {
  async byBundleIdentifierAsync(
    graphqlClient: ExpoGraphqlClient,
    accountName: string,
    bundleIdentifier: string
  ): Promise<AppleAppIdentifierFragment | null> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<AppleAppIdentifierByBundleIdQuery>(
          gql`
            query AppleAppIdentifierByBundleIdQuery(
              $accountName: String!
              $bundleIdentifier: String!
            ) {
              account {
                byName(accountName: $accountName) {
                  id
                  appleAppIdentifiers(bundleIdentifier: $bundleIdentifier) {
                    id
                    ...AppleAppIdentifierFragment
                  }
                }
              }
            }
            ${print(AppleAppIdentifierFragmentNode)}
          `,
          {
            accountName,
            bundleIdentifier,
          },
          {
            additionalTypenames: ['AppleAppIdentifier'],
          }
        )
        .toPromise()
    );
    return data.account.byName.appleAppIdentifiers[0];
  },
};
