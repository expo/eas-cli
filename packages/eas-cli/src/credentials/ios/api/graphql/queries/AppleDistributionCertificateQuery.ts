import assert from 'assert';
import { print } from 'graphql';
import gql from 'graphql-tag';

import { graphqlClient, withErrorHandlingAsync } from '../../../../../graphql/client';
import {
  AppleDistributionCertificateByAccountQuery,
  AppleDistributionCertificateByAppQuery,
  AppleDistributionCertificateFragment,
  IosDistributionType,
} from '../../../../../graphql/generated';
import { AppleDistributionCertificateFragmentNode } from '../../../../../graphql/types/credentials/AppleDistributionCertificate';
import { AppleTeamFragmentNode } from '../../../../../graphql/types/credentials/AppleTeam';

const AppleDistributionCertificateQuery = {
  async getForAppAsync(
    projectFullName: string,
    {
      appleAppIdentifierId,
      iosDistributionType,
    }: { appleAppIdentifierId: string; iosDistributionType: IosDistributionType }
  ): Promise<AppleDistributionCertificateFragment | null> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<AppleDistributionCertificateByAppQuery>(
          gql`
            query AppleDistributionCertificateByAppQuery(
              $projectFullName: String!
              $appleAppIdentifierId: String!
              $iosDistributionType: IosDistributionType!
            ) {
              app {
                byFullName(fullName: $projectFullName) {
                  id
                  iosAppCredentials(filter: { appleAppIdentifierId: $appleAppIdentifierId }) {
                    id
                    iosAppBuildCredentialsArray(
                      filter: { iosDistributionType: $iosDistributionType }
                    ) {
                      id
                      distributionCertificate {
                        id
                        ...AppleDistributionCertificateFragment
                        appleTeam {
                          id
                          ...AppleTeamFragment
                        }
                      }
                    }
                  }
                }
              }
            }
            ${print(AppleDistributionCertificateFragmentNode)}
            ${print(AppleTeamFragmentNode)}
          `,
          {
            projectFullName,
            appleAppIdentifierId,
            iosDistributionType,
          },
          {
            additionalTypenames: ['IosAppCredentials', 'IosAppBuildCredentials'],
          }
        )
        .toPromise()
    );
    assert(data.app, 'GraphQL: `app` not defined in server response');
    return (
      data.app.byFullName.iosAppCredentials[0]?.iosAppBuildCredentialsArray[0]
        ?.distributionCertificate ?? null
    );
  },
  async getAllForAccount(accountName: string): Promise<AppleDistributionCertificateFragment[]> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<AppleDistributionCertificateByAccountQuery>(
          gql`
            query AppleDistributionCertificateByAccountQuery($accountName: String!) {
              account {
                byName(accountName: $accountName) {
                  id
                  appleDistributionCertificates {
                    id
                    ...AppleDistributionCertificateFragment
                  }
                }
              }
            }
            ${print(AppleDistributionCertificateFragmentNode)}
          `,
          {
            accountName,
          }
        )
        .toPromise()
    );
    return data.account.byName.appleDistributionCertificates;
  },
};

export { AppleDistributionCertificateQuery };
