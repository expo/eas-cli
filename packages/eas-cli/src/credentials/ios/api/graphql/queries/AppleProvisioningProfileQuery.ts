import assert from 'assert';
import { print } from 'graphql';
import gql from 'graphql-tag';

import { ExpoGraphqlClient } from '../../../../../commandUtils/context/contextUtils/createGraphqlClient';
import { withErrorHandlingAsync } from '../../../../../graphql/client';
import {
  AppleAppIdentifierFragment,
  AppleDeviceFragment,
  AppleProvisioningProfileFragment,
  AppleProvisioningProfilesByAppQuery,
  AppleTeamFragment,
  IosDistributionType,
} from '../../../../../graphql/generated';
import { AppleAppIdentifierFragmentNode } from '../../../../../graphql/types/credentials/AppleAppIdentifier';
import { AppleDeviceFragmentNode } from '../../../../../graphql/types/credentials/AppleDevice';
import { AppleProvisioningProfileFragmentNode } from '../../../../../graphql/types/credentials/AppleProvisioningProfile';
import { AppleTeamFragmentNode } from '../../../../../graphql/types/credentials/AppleTeam';

export type AppleProvisioningProfileQueryResult = AppleProvisioningProfileFragment & {
  appleTeam?: AppleTeamFragment | null;
} & { appleDevices: AppleDeviceFragment[] } & { appleAppIdentifier: AppleAppIdentifierFragment };
export const AppleProvisioningProfileQuery = {
  async getForAppAsync(
    graphqlClient: ExpoGraphqlClient,
    projectFullName: string,
    {
      appleAppIdentifierId,
      iosDistributionType,
    }: { appleAppIdentifierId: string; iosDistributionType: IosDistributionType }
  ): Promise<AppleProvisioningProfileQueryResult | null> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<AppleProvisioningProfilesByAppQuery>(
          gql`
            query AppleProvisioningProfilesByAppQuery(
              $projectFullName: String!
              $appleAppIdentifierId: String!
              $iosDistributionType: IosDistributionType!
            ) {
              app {
                byFullName(fullName: $projectFullName) {
                  id
                  iosAppCredentials(filter: { appleAppIdentifierId: $appleAppIdentifierId }) {
                    id
                    iosAppBuildCredentialsList(
                      filter: { iosDistributionType: $iosDistributionType }
                    ) {
                      id
                      provisioningProfile {
                        id
                        ...AppleProvisioningProfileFragment
                        appleTeam {
                          id
                          ...AppleTeamFragment
                        }
                        appleDevices {
                          id
                          ...AppleDeviceFragment
                        }
                        appleAppIdentifier {
                          id
                          ...AppleAppIdentifierFragment
                        }
                      }
                    }
                  }
                }
              }
            }
            ${print(AppleProvisioningProfileFragmentNode)}
            ${print(AppleTeamFragmentNode)}
            ${print(AppleDeviceFragmentNode)}
            ${print(AppleAppIdentifierFragmentNode)}
          `,
          {
            projectFullName,
            appleAppIdentifierId,
            iosDistributionType,
          },
          {
            additionalTypenames: [
              'AppleProvisioningProfile',
              'IosAppCredentials',
              'IosAppBuildCredentials',
            ],
          }
        )
        .toPromise()
    );
    assert(data.app, 'GraphQL: `app` not defined in server response');
    return (
      data.app.byFullName.iosAppCredentials[0]?.iosAppBuildCredentialsList[0]
        ?.provisioningProfile ?? null
    );
  },
};
