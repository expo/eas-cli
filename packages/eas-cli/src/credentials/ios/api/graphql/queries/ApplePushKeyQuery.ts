import { print } from 'graphql';
import gql from 'graphql-tag';

import { ExpoGraphqlClient } from '../../../../../commandUtils/context/contextUtils/createGraphqlClient';
import { withErrorHandlingAsync } from '../../../../../graphql/client';
import { ApplePushKeyByAccountQuery, ApplePushKeyFragment } from '../../../../../graphql/generated';
import { ApplePushKeyFragmentNode } from '../../../../../graphql/types/credentials/ApplePushKey';

export const ApplePushKeyQuery = {
  async getAllForAccountAsync(
    graphqlClient: ExpoGraphqlClient,
    accountName: string
  ): Promise<ApplePushKeyFragment[]> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<ApplePushKeyByAccountQuery>(
          gql`
            query ApplePushKeyByAccountQuery($accountName: String!) {
              account {
                byName(accountName: $accountName) {
                  id
                  applePushKeys {
                    id
                    ...ApplePushKeyFragment
                  }
                }
              }
            }
            ${print(ApplePushKeyFragmentNode)}
          `,
          {
            accountName,
          },
          {
            additionalTypenames: ['ApplePushKey'],
          }
        )
        .toPromise()
    );
    return data.account.byName.applePushKeys;
  },
};
