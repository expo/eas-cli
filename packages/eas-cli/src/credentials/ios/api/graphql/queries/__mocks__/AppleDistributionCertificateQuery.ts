export const AppleDistributionCertificateQuery = {
  getAllForAccountAsync: jest.fn().mockImplementation(() => {
    return [
      {
        id: '6a7422cc-392b-451d-984c-b0b4b027680a',
        certificateP12: 'test-cert-p12-1',
        certificatePassword: 'test-cert-password-1',
        serialNumber: '5544BE191B9F949E',
        developerPortalIdentifier: 'JRH7292L8R',
        validityNotBefore: '2020-06-22T20:46:07.000Z',
        validityNotAfter: '2021-06-22T20:46:07.000Z',
        appleTeam: {
          id: '6d39dbfe-a30a-44a4-bc36-d562e5963fa3',
          appleTeamIdentifier: '77KQ969CHE',
          appleTeamName: null,
          __typename: 'AppleTeam',
        },
        iosAppBuildCredentialsList: [],
        updatedAt: '2020-06-22T20:46:07.000Z',
        __typename: 'AppleDistributionCertificate',
      },
      {
        id: '8b3baa53-b149-4a1b-842b-75ff6e0bcb1e',
        certificateP12: 'test-cert-p12-2',
        certificatePassword: 'test-cert-password-2',
        serialNumber: '3F9CDD6CFC6E52C1',
        developerPortalIdentifier: '5URSS525PL',
        validityNotBefore: '2020-10-13T18:28:44.000Z',
        validityNotAfter: '2021-10-13T18:28:44.000Z',
        appleTeam: {
          id: '6d39dbfe-a30a-44a4-bc36-d562e5963fa3',
          appleTeamIdentifier: '77KQ969CHE',
          appleTeamName: null,
          __typename: 'AppleTeam',
        },
        iosAppBuildCredentialsList: [],
        updatedAt: '2020-06-22T20:46:07.000Z',
        __typename: 'AppleDistributionCertificate',
      },
    ];
  }),
};
