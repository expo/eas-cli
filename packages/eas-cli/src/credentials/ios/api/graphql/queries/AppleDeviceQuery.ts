import assert from 'assert';
import { print } from 'graphql';
import gql from 'graphql-tag';

import { ExpoGraphqlClient } from '../../../../../commandUtils/context/contextUtils/createGraphqlClient';
import { DeviceNotFoundError } from '../../../../../devices/utils/errors';
import { withErrorHandlingAsync } from '../../../../../graphql/client';
import {
  AppleDevice,
  AppleDeviceFragment,
  AppleDevicesByAppleTeamQuery,
  AppleDevicesByIdentifierQuery,
  AppleDevicesByTeamIdentifierQuery,
  AppleDevicesByTeamIdentifierQueryVariables,
  AppleTeamFragment,
} from '../../../../../graphql/generated';
import { AppleDeviceFragmentNode } from '../../../../../graphql/types/credentials/AppleDevice';
import { AppleTeamFragmentNode } from '../../../../../graphql/types/credentials/AppleTeam';
import { formatAppleTeam } from '../../../actions/AppleTeamFormatting';

export type AppleDeviceFragmentWithAppleTeam = AppleDeviceFragment & {
  appleTeam: AppleTeamFragment;
};

export type AppleDeviceQueryResult = Pick<
  AppleDevice,
  'id' | 'identifier' | 'name' | 'deviceClass' | 'enabled'
>;

export type AppleDevicesByTeamIdentifierQueryResult = AppleTeamFragment & {
  appleDevices: AppleDeviceQueryResult[];
};

export type AppleDevicesByIdentifierQueryResult = AppleDeviceQueryResult & {
  appleTeam: AppleTeamFragment;
};

export const AppleDeviceQuery = {
  async getAllByAppleTeamIdentifierAsync(
    graphqlClient: ExpoGraphqlClient,
    accountId: string,
    appleTeamIdentifier: string,
    { useCache = true }: { useCache?: boolean } = {}
  ): Promise<AppleDeviceFragmentWithAppleTeam[]> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<AppleDevicesByAppleTeamQuery>(
          gql`
            query AppleDevicesByAppleTeamQuery($accountId: ID!, $appleTeamIdentifier: String!) {
              appleTeam {
                byAppleTeamIdentifier(accountId: $accountId, identifier: $appleTeamIdentifier) {
                  id
                  ...AppleTeamFragment
                  appleDevices {
                    id
                    ...AppleDeviceFragment
                    appleTeam {
                      id
                      ...AppleTeamFragment
                    }
                  }
                }
              }
            }
            ${print(AppleTeamFragmentNode)}
            ${print(AppleDeviceFragmentNode)}
          `,
          {
            accountId,
            appleTeamIdentifier,
          },
          {
            additionalTypenames: ['AppleDevice'],
            requestPolicy: useCache ? 'cache-first' : 'network-only',
          }
        )
        .toPromise()
    );
    assert(
      data.appleTeam.byAppleTeamIdentifier,
      'byAppleTeamIdentifier should be defined in this context - enforced by GraphQL'
    );
    const { appleDevices } = data.appleTeam.byAppleTeamIdentifier;
    assert(appleDevices, 'Apple Devices should be defined in this context - enforced by GraphQL');
    return appleDevices;
  },

  async getAllForAppleTeamAsync(
    graphqlClient: ExpoGraphqlClient,
    { accountName, appleTeamIdentifier, offset, limit }: AppleDevicesByTeamIdentifierQueryVariables
  ): Promise<AppleDeviceFragment[]> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<AppleDevicesByTeamIdentifierQuery>(
          gql`
            query AppleDevicesByTeamIdentifier(
              $accountName: String!
              $appleTeamIdentifier: String!
              $offset: Int
              $limit: Int
            ) {
              account {
                byName(accountName: $accountName) {
                  id
                  appleTeams(appleTeamIdentifier: $appleTeamIdentifier) {
                    id
                    appleTeamIdentifier
                    appleTeamName
                    appleDevices(offset: $offset, limit: $limit) {
                      id
                      identifier
                      name
                      deviceClass
                      enabled
                      model
                    }
                  }
                }
              }
            }
          `,
          { accountName, appleTeamIdentifier, offset, limit },
          {
            additionalTypenames: ['AppleDevice', 'AppleTeam'],
          }
        )
        .toPromise()
    );
    const [appleTeam] = data.account.byName.appleTeams;
    const { appleDevices } = appleTeam;
    if (!appleDevices) {
      throw new Error(`Could not find devices on Apple team -- ${formatAppleTeam(appleTeam)}`);
    }
    return appleDevices;
  },

  async getByDeviceIdentifierAsync(
    graphqlClient: ExpoGraphqlClient,
    accountName: string,
    identifier: string
  ): Promise<AppleDevicesByIdentifierQueryResult> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .query<AppleDevicesByIdentifierQuery>(
          gql`
            query AppleDevicesByIdentifier($accountName: String!, $identifier: String!) {
              account {
                byName(accountName: $accountName) {
                  id
                  appleDevices(identifier: $identifier) {
                    id
                    model
                    identifier
                    name
                    deviceClass
                    enabled
                    appleTeam {
                      id
                      appleTeamIdentifier
                      appleTeamName
                    }
                  }
                }
              }
            }
          `,
          { accountName, identifier },
          {
            additionalTypenames: ['AppleDevice', 'AppleTeam'],
          }
        )
        .toPromise()
    );

    const device = data.account.byName.appleDevices[0];
    if (!device) {
      throw new DeviceNotFoundError(
        `Device with id ${identifier} was not found on account ${accountName}.`
      );
    }
    return device;
  },
};
