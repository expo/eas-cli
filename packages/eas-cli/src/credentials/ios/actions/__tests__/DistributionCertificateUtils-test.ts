import mockdate from 'mockdate';
import { instance, mock } from 'ts-mockito';

import { ExpoGraphqlClient } from '../../../../commandUtils/context/contextUtils/createGraphqlClient';
import { AppleDistributionCertificateQuery } from '../../api/graphql/queries/AppleDistributionCertificateQuery';
import { formatDistributionCertificate } from '../DistributionCertificateUtils';
jest.mock('../../api/graphql/queries/AppleDistributionCertificateQuery');
jest.mock('chalk', () => {
  return {
    __esModule: true, // this property makes it work
    default: { gray: jest.fn(log => log), underline: jest.fn(log => log) },
  };
});

mockdate.set(new Date('4/20/2021'));
describe('select credentials', () => {
  it('select an AppleDistributionCertificate fragment', async () => {
    const graphqlClient = instance(mock<ExpoGraphqlClient>());
    const testDistCerts = (
      await AppleDistributionCertificateQuery.getAllForAccountAsync(graphqlClient, 'quinAccount')
    ).sort((a, b) => (a.serialNumber > b.serialNumber ? 1 : -1));
    const loggedSoFar = testDistCerts
      .map(cert => formatDistributionCertificate(cert))
      .reduce((acc, certLog) => acc + certLog);
    expect(loggedSoFar).toMatchSnapshot();
  });
});
