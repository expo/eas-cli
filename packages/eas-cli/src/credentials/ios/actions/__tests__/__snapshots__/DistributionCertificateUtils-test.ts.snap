// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select credentials select an AppleDistributionCertificate fragment 1`] = `
"Cert ID: 5URSS525PL, Serial number: 3F9CDD6CFC6E52C1, Team ID: 77KQ969CHE
    Created: 6 months ago, Updated: 10 months ago,
    Expires: Wed, 13 Oct 2021 18:28:44 UTCCert ID: JRH7292L8R, Serial number: 5544BE191B9F949E, Team ID: 77KQ969CHE
    Created: 10 months ago, Updated: 10 months ago,
    Expires: Tue, 22 Jun 2021 20:46:07 UTC"
`;
